File: base44.js
/**
 * Base44 Encoding/Decoding for Browser
 * Browser-compatible version using Uint8Array instead of Node.js Buffer
 */

const Base44 = {
    ALPHABET: "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefgh",
    BASE: 44n,

    /**
     * Encode Uint8Array to Base44 string
     */
    encode(data) {
        if (!data || data.length === 0) {
            return "";
        }

        // Convert bytes to BigInt
        let num = 0n;
        for (let i = 0; i < data.length; i++) {
            num = (num << 8n) | BigInt(data[i]);
        }

        // Convert to base44
        if (num === 0n) {
            return this.ALPHABET[0];
        }

        const result = [];
        while (num > 0n) {
            const remainder = Number(num % this.BASE);
            result.push(this.ALPHABET[remainder]);
            num = num / this.BASE;
        }

        result.reverse();

        // Handle leading zeros
        let leadingZeros = 0;
        for (let i = 0; i < data.length; i++) {
            if (data[i] === 0) {
                leadingZeros++;
            } else {
                break;
            }
        }

        return this.ALPHABET[0].repeat(leadingZeros) + result.join('');
    },

    /**
     * Decode Base44 string to Uint8Array
     */
    decode(encoded) {
        if (!encoded || encoded.length === 0) {
            return new Uint8Array(0);
        }

        // Count leading zeros
        let leadingZeros = 0;
        for (let i = 0; i < encoded.length; i++) {
            if (encoded[i] === this.ALPHABET[0]) {
                leadingZeros++;
            } else {
                break;
            }
        }

        // Convert from base44 to BigInt
        let num = 0n;
        for (let i = 0; i < encoded.length; i++) {
            const char = encoded[i];
            const index = this.ALPHABET.indexOf(char);
            if (index === -1) {
                throw new Error(`Invalid character in Base44 string: ${char}`);
            }
            num = num * this.BASE + BigInt(index);
        }

        // Convert BigInt to bytes
        if (num === 0n) {
            return new Uint8Array(Math.max(leadingZeros, 1));
        }

        // Convert to byte array
        const bytes = [];
        let tempNum = num;
        while (tempNum > 0n) {
            bytes.unshift(Number(tempNum & 0xFFn));
            tempNum = tempNum >> 8n;
        }

        // Create Uint8Array with leading zeros
        const result = new Uint8Array(leadingZeros + bytes.length);
        for (let i = 0; i < bytes.length; i++) {
            result[leadingZeros + i] = bytes[i];
        }

        return result;
    }
};

File: base44.py
"""
Base44 Encoding/Decoding Module

Implements Base44 encoding using a 44-character alphabet:
A-Z (26 chars), a-z (26 chars, but we only use first 8), 0-9 (10 chars) = 44 chars total
Alphabet: ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefgh-_
"""


class Base44:
    """Base44 encoder/decoder using 44-character alphabet"""
    
    # First 44 characters: A-Z (26), 0-9 (10), a-h (8) = 44 chars
    ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789abcdefgh-_"[:44]
    BASE = 44
    
    @classmethod
    def encode(cls, data: bytes) -> str:
        """
        Encode bytes to Base44 string
        
        Args:
            data: Bytes to encode
            
        Returns:
            Base44 encoded string
        """
        if not data:
            return ""
        
        # Convert bytes to integer
        num = int.from_bytes(data, byteorder='big')
        
        # Convert to base44
        if num == 0:
            return cls.ALPHABET[0]
        
        result = []
        while num > 0:
            num, remainder = divmod(num, cls.BASE)
            result.append(cls.ALPHABET[remainder])
        
        # Reverse to get correct order
        result.reverse()
        
        # Handle leading zeros
        leading_zeros = len(data) - len(data.lstrip(b'\x00'))
        return cls.ALPHABET[0] * leading_zeros + ''.join(result)
    
    @classmethod
    def decode(cls, encoded: str) -> bytes:
        """
        Decode Base44 string to bytes
        
        Args:
            encoded: Base44 encoded string
            
        Returns:
            Decoded bytes
        """
        if not encoded:
            return b""
        
        # Count leading zeros
        leading_zeros = len(encoded) - len(encoded.lstrip(cls.ALPHABET[0]))
        
        # Convert from base44 to integer
        num = 0
        for char in encoded:
            if char not in cls.ALPHABET:
                raise ValueError(f"Invalid character in Base44 string: {char}")
            num = num * cls.BASE + cls.ALPHABET.index(char)
        
        # Convert integer to bytes
        if num == 0:
            # If the entire string is just zeros, return the correct number of zero bytes
            return b'\x00' * max(leading_zeros, 1)
        
        byte_length = (num.bit_length() + 7) // 8
        result = num.to_bytes(byte_length, byteorder='big')
        
        # Add leading zero bytes
        return b'\x00' * leading_zeros + result


def encode(data: bytes) -> str:
    """Encode bytes to Base44 string"""
    return Base44.encode(data)


def decode(encoded: str) -> bytes:
    """Decode Base44 string to bytes"""
    return Base44.decode(encoded)


if __name__ == "__main__":
    print("=" * 60)
    print("Base44 Encoding/Decoding Test")
    print("=" * 60)
    
    # Test cases
    test_cases = [
        b"Hello, World!",
        "Kurdish: Silav, tu çawa yî?".encode('utf-8'),
        b"German: Guten Tag!",
        b"French: Bonjour!",
        b"Turkish: Merhaba!",
        b"English: Hello!",
        b"\x00\x00\x01\x02",  # Test leading zeros
        b"",  # Empty
        b"\x00",  # Single zero
        bytes(range(256)),  # All possible bytes
    ]
    
    print("\n✅ Running test cases...\n")
    
    all_passed = True
    for i, test_data in enumerate(test_cases, 1):
        try:
            # Encode
            encoded = encode(test_data)
            
            # Decode
            decoded = decode(encoded)
            
            # Verify
            if decoded == test_data:
                status = "✅ PASS"
                if len(test_data) <= 50:
                    print(f"Test {i}: {status}")
                    print(f"  Original: {test_data[:50]}")
                    print(f"  Encoded:  {encoded[:80]}{'...' if len(encoded) > 80 else ''}")
                    print(f"  Size:     {len(test_data)} bytes → {len(encoded)} chars")
                    print(f"  Ratio:    {len(encoded) / max(len(test_data), 1):.2f}x")
                else:
                    print(f"Test {i}: {status} (large data: {len(test_data)} bytes)")
                    print(f"  Encoded size: {len(encoded)} chars")
                    print(f"  Ratio: {len(encoded) / len(test_data):.2f}x")
            else:
                status = "❌ FAIL"
                all_passed = False
                print(f"Test {i}: {status}")
                print(f"  Original: {test_data}")
                print(f"  Decoded:  {decoded}")
        except Exception as e:
            all_passed = False
            print(f"Test {i}: ❌ ERROR - {e}")
        
        print()
    
    print("=" * 60)
    if all_passed:
        print("✅ All tests passed!")
    else:
        print("❌ Some tests failed!")
    print("=" * 60)
